#! /bin/sh -e

# DP: Fix PR other/28322, GCC new warnings and compatibility.

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

gcc/

2008-02-26  Manuel Lopez-Ibanez <manu@gcc.gnu.org>

	PR other/28322
	* toplev.c (toplev_main): If there are warnings or error, print
	errors for ignored options.
	* opts.c (ignored_options): New static variable.
	(postpone_unknown_option_error): New.
	(print_ignored_options): New.
	(handle_option): Postpone errors for unknown -Wno-* options.
	* opts.h (print_ignored_options): Declare.

2008-06-03  Peter Maydell  <pmaydell@chiark.greenend.org.uk>

	PR other/28322
	* opts.c (print_ignored_options): report postponed diagnostics for
	unknown	-Wno-* options as warnings, not errors.
	(postpone_unknown_option_error): renamed to
	postpone_unknown_option_warning.

gcc/testsuite/

2008-02-26  Manuel Lopez-Ibanez  <manu@gcc.gnu.org>

	PR other/28322
	* gcc.dg/pr28322.c: New.
	* gcc.dg/pr28322-2.c: New.
	* lib/prune.exp: Ignore "At top level" even if there is no ':'
	preceding it.
	* gcc.dg/pr28322-3.c: New.

2008-06-03  Peter Maydell  <pmaydell@chiark.greenend.org.uk>

	* gcc.dg/pr28322-2.c: check that emitted diagnostic for -Wno-foobar
	is a warning and not an error.


--- gcc/opts.c.orig	2008-03-08 13:32:39.000000000 +0000
+++ gcc/opts.c	2009-01-17 16:48:14.000000000 +0000
@@ -97,6 +97,12 @@
 static bool flag_value_profile_transformations_set;
 static bool flag_peel_loops_set, flag_branch_probabilities_set;
 
+typedef const char *const_char_p; /* For DEF_VEC_P.  */
+DEF_VEC_P(const_char_p);
+DEF_VEC_ALLOC_P(const_char_p,heap);
+
+static VEC(const_char_p,heap) *ignored_options;
+
 /* Input file names.  */
 const char **in_fnames;
 unsigned num_in_fnames;
@@ -181,6 +187,36 @@
   free (bad_lang);
 }
 
+/* Buffer the unknown option described by the string OPT.  Currently,
+   we only complain about unknown -Wno-* options if they may have
+   prevented a diagnostic. Otherwise, we just ignore them.
+   Note that if we do complain, it is only as a warning, not an error;
+   passing the compiler an unrecognised -Wno-* option should never
+   change whether the compilation succeeds or fails.  */
+
+static void postpone_unknown_option_warning(const char *opt)
+{
+  VEC_safe_push (const_char_p, heap, ignored_options, opt);
+}
+
+/* Produce a warning for each option previously buffered.  */
+
+void print_ignored_options (void)
+{
+  location_t saved_loc = input_location;
+
+  input_location = UNKNOWN_LOCATION;
+
+  while (!VEC_empty (const_char_p, ignored_options))
+    {
+      const char *opt;
+      opt = VEC_pop (const_char_p, ignored_options);
+      warning (0, "unrecognized command line option \"%s\"", opt);
+    }
+
+  input_location = saved_loc;
+}
+
 /* Handle the switch beginning at ARGV for the language indicated by
    LANG_MASK.  Returns the number of switches consumed.  */
 static unsigned int
@@ -210,6 +246,14 @@
       opt = dup;
       value = 0;
       opt_index = find_opt (opt + 1, lang_mask | CL_COMMON | CL_TARGET);
+      if (opt_index == cl_options_count && opt[1] == 'W')
+	{
+	  /* We don't generate warnings for unknown -Wno-* options
+             unless we issue diagnostics.  */
+	  postpone_unknown_option_warning (argv[0]);
+	  result = 1;
+	  goto done;
+	}
     }
 
   if (opt_index == cl_options_count)
--- gcc/opts.h.orig	2007-09-03 11:05:59.000000000 +0000
+++ gcc/opts.h	2009-01-17 16:47:22.000000000 +0000
@@ -91,4 +91,5 @@
 extern int option_enabled (int opt_idx);
 extern bool get_option_state (int, struct cl_option_state *);
 
+extern void print_ignored_options (void);
 #endif
--- gcc/toplev.c.orig	2009-01-17 16:46:12.000000000 +0000
+++ gcc/toplev.c	2009-01-17 16:47:22.000000000 +0000
@@ -2034,6 +2034,9 @@
   if (!exit_after_options)
     do_compile ();
 
+  if (warningcount || errorcount) 
+    print_ignored_options ();
+
   if (errorcount || sorrycount)
     return (FATAL_EXIT_CODE);
 
Index: gcc/testsuite/gcc.dg/pr28322-2.c
===================================================================
--- gcc/testsuite/gcc.dg/pr28322-2.c	(revision 0)
+++ gcc/testsuite/gcc.dg/pr28322-2.c	(revision 0)
@@ -0,0 +1,24 @@
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options "-Wall -Wextra -Wno-foobar" } */
+
+int foo (void) 
+{
+  int i = 1/0;  /* { dg-warning "division by zero" } */
+  return i;
+}
+
+/* { dg-message "unrecognized command line option .-Wno-foobar." "" { target *-*-* } 0 } */
+
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options "-Wall -Wextra -Wno-foobar" } */
+
+int foo (void) 
+{
+  int i = 1/0;  /* { dg-warning "division by zero" } */
+  return i;
+}
+
+/* { dg-warning "unrecognized command line option .-Wno-foobar." "" { target *-*-* } 0 } */
+
Index: gcc/testsuite/gcc.dg/pr28322.c
===================================================================
--- gcc/testsuite/gcc.dg/pr28322.c	(revision 0)
+++ gcc/testsuite/gcc.dg/pr28322.c	(revision 0)
@@ -0,0 +1,16 @@
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options " -Wno-foobar -Wno-div-by-zero" } */
+
+void foo(void)
+{
+  int i =  1/0;
+}
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options " -Wno-foobar -Wno-div-by-zero" } */
+
+void foo(void)
+{
+  int i =  1/0;
+}
Index: gcc/testsuite/gcc.dg/pr28322-3.c
===================================================================
--- gcc/testsuite/gcc.dg/pr28322-3.c	(revision 0)
+++ gcc/testsuite/gcc.dg/pr28322-3.c	(revision 0)
@@ -0,0 +1,20 @@
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options " -fno-foobar -mno-foobar" } */
+
+void foo(void)
+{
+  int i =  1;
+}
+/* { dg-message "unrecognized command line option .-fno-foobar." "" { target *-*-* } 0 } */
+/* { dg-message "unrecognized command line option .-mno-foobar." "" { target *-*-* } 0 } */
+/* PR28322: ignore unknown -Wno-* if no warning is emitted.  */
+/* { dg-do compile } */
+/* { dg-options " -fno-foobar -mno-foobar" } */
+
+void foo(void)
+{
+  int i =  1;
+}
+/* { dg-warning "unrecognized command line option .-fno-foobar." "" { target *-*-* } 0 } */
+/* { dg-warning "unrecognized command line option .-mno-foobar." "" { target *-*-* } 0 } */
Index: gcc/testsuite/lib/prune.exp
===================================================================
--- gcc/testsuite/lib/prune.exp	(revision 134648)
+++ gcc/testsuite/lib/prune.exp	(working copy)
@@ -21,7 +21,7 @@
     #send_user "Before:$text\n"
 
     regsub -all "(^|\n)\[^\n\]*: In ((static member )?function|member|method|(copy )?constructor|destructor|instantiation|program|subroutine|block-data) \[^\n\]*" $text "" text
-    regsub -all "(^|\n)\[^\n\]*: At (top level|global scope):\[^\n\]*" $text "" text
+    regsub -all "(^|\n)\[^\n\]*(: )?At (top level|global scope):\[^\n\]*" $text "" text
     regsub -all "(^|\n)\[^\n\]*:   instantiated from \[^\n\]*" $text "" text
     regsub -all "(^|\n)collect2: ld returned \[^\n\]*" $text "" text
     regsub -all "(^|\n)collect: re(compiling|linking)\[^\n\]*" $text "" text
